lib/wallet/src/Cardano/Wallet.hs

```diff
@@ -2189,10 +2210,53 @@ balanceTx wrk pp timeTranslation partialTx = do
         utxoIndex
         (defaultChangeAddressGen argGenChange)
         changeState
-        partialTx
+        (over #extraUTxO (<> lookedUpUTxO) partialTx)
 
     return tx
   where
+    era = Write.recentEra @era
+
+    -- | 'assumeUTxOInEra era utxo' will convert the 'utxo' to the given era,
+    -- and throw an exception in 'IO' if the conversion fails.
+    --
+    -- It assumes the 'utxo' was queried from the node /after/ the 'era'.
+    assumeUTxOInEra
+        :: RecentEra era
+        -> MaybeInRecentEra Write.UTxO
+        -> IO (Write.UTxO era)
+    assumeUTxOInEra RecentEraConway (InRecentEraConway u)
+        = pure u
+    assumeUTxOInEra RecentEraBabbage (InRecentEraBabbage u)
+        = pure u
+    assumeUTxOInEra RecentEraBabbage (InRecentEraConway u)
+        -- The era changed to Conway between fetching the era and fetching the
+        -- utxo. Let's downgrade the utxo and continue.
+        = either (throwIO . ExceptionInvalidTxOutInEra) pure $ downgradeEra u
+    assumeUTxOInEra RecentEraConway (InRecentEraBabbage _)
+        -- It should be impossible for the node's tip to change from Conway back
+        -- to Babbage.
+        = impossibleRollback
```

Comment:
Should be impossible, but it is also very easy to upgrade the TxOuts. Think I should do that instead.